fix(ngModel): fix issues when parserName is same as validator key

Narretz opened the pull request:
For $validate(), it is necessary to store the parseError state
in the controller. Otherwise, if the parser name equals a validator
key, $validate() will assume a parse error occured if the validator
is invalid.

Also, setting the validity for the parser now happens after setting
validity for the validator key. Otherwise, the parse key is set,
and then immediately afterwards the validator key is unset
(because parse errors remove all other validations).

Fixes #10698
Closes #10828

Narretz committed:
fix(ngModel): fix issues when parserName is same as validator key

For $validate(), it is necessary to store the parseError state
in the controller. Otherwise, if the parser name equals a validator
key, $validate() will assume a parse error occured if the validator
is invalid.

Also, setting the validity for the parser now happens after setting
validity for the validator key. Otherwise, the parse key is set,
and then immediately afterwards the validator key is unset
(because parse errors remove all other validations).

Fixes #10698
Closes #10850

petebacondarwin commented on src/ng/directive/ngModel.js:
```diff
@@ -518,7 +519,7 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
 
     // Check if the there's a parse error, so we don't unset it accidentially
     var parserName = ctrl.$$parserName || 'parse';
```
This variable doesn't seem to be used now.  What does this mean for the default of 'parse'?

petebacondarwin replied:
I see that the `errorKey` variable sets the default below on line 563

petebacondarwin commented on src/ng/directive/ngModel.js:
```diff
@@ -571,9 +571,14 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
-          return false;
+          ctrl.$$parseError = true;
         }
+        // Set the parse error last, to prevent unsetting it, should a $validators key == parserName
```
This comment doesn't read very well. Also if we remove the unused `parserName` variable then this makes less sense.

petebacondarwin commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -1221,6 +1221,80 @@ describe('ngModel', function() {
+          return otherValid;
+        };
+
+        ctrl.$$parserName = 'test';
```
I think `ctrl.$$parserName` was a bad idea in the first place. We ought to refactor this in general if possible.

Narretz replied:
Any ideas? It's currently a private API, but we rely on it for the name of the error key. A separate function, $setParseError(name) might be possible that is called inside the parse error. Although I think we have to stick with `return undefined` denoting a parse error, bad as it is.

Narretz replied:
We could also allow returning an object that contains parser name, error state, and return value. This would allow returning undefined as a valid value, as well as get rid of $$parserName and expose an API to actually set the parser name.

petebacondarwin replied:
How about we use the `name` on the parser function? This would let us provide a named function for a parse and get a free name, or we can specify it explicitly.

Narretz replied:
I just noticed that `name` is not supported in IE9+ - we'd have to polyfill it.

petebacondarwin replied:
Hmm, how about we do something like:  `parser.name || parser.$$name` or something to allow people to specify the name explicitly if they need to support IE9

Narretz replied:
It's all IEs, up to IE11, so I don't think it's an option.

petebacondarwin commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -1221,6 +1221,80 @@ describe('ngModel', function() {
+        expect(scope.value).toBeUndefined();
+        expect(ctrl.$error).toEqual({test: true});
+      });
+
```
This test seems overly complicated. Couldn't we replace the parser and validators with simple functions, like:

* parser -> `isEven(value)`
* validator1 -> `isOdd(value)`
* validator2 -> `isGreaterThanTen(value)`

Then we don't have to grok as many variables - just simply send in the following values:

* `1`
* `2`
* `11`
* `12`

And check that we get the correct validation?

Narretz committed:
test(ngModel): refactor the test

Narretz commented:
I updated the test. I removed the variables that handle the validity of parser / validators. Instead, they now directly react to the input.
I didn't use the even / odd / largerThanTen scheme, because the parser and one validator need to have the same name, but would have validated different things. This is imo even more confusing. Now it's very verbose, but also quite clear what is tested.

If you're okay with the test, can we merge this, and push the parseName refactor back? The bug is pretty gnarly and we can fix the parser name later.

petebacondarwin commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -1221,6 +1221,96 @@ describe('ngModel', function() {
         expect(ctrl.$validators.mock.calls.length).toEqual(2);
       });
+
+      iit('should validate correctly when $parser name equals $validator key', function() {
```
iit

petebacondarwin commented:
Oh! Sorry @Narretz - when I squashed I attributed it to me instead of you. I am sorry about that - I will not do it again. Great work.
